File: auto-upload.py
"""
Auto-Upload Product Images to Supabase
Watches the 'product-images' folder and automatically uploads new images
"""

import os
import time
import json
from pathlib import Path
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler
import requests
from datetime import datetime

# Supabase Configuration
SUPABASE_URL = "https://nzwtafacdpdgulzcwntx.supabase.co"
SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im56d3RhZmFjZHBkZ3VsemN3bnR4Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjU4MjkxNDksImV4cCI6MjA4MTQwNTE0OX0.6elrAvcsEAc0Jaj4P8-ZFLSWKi2cvzgoAYGlDxeR-8U"

# Paths
SCRIPT_DIR = Path(__file__).parent
IMAGES_FOLDER = SCRIPT_DIR / "product-images"
PROCESSED_FILE = SCRIPT_DIR / ".processed-images.json"

# Supported image formats
SUPPORTED_FORMATS = {'.jpg', '.jpeg', '.png', '.gif', '.webp'}

class ProcessedImages:
    def __init__(self):
        self.data = self.load()
    
    def load(self):
        if PROCESSED_FILE.exists():
            try:
                with open(PROCESSED_FILE, 'r') as f:
                    return json.load(f)
            except:
                return {}
        return {}
    
    def save(self):
        with open(PROCESSED_FILE, 'w') as f:
            json.dump(self.data, f, indent=2)
    
    def is_processed(self, filename):
        return filename in self.data
    
    def mark_processed(self, filename, product_id, storage_path, url):
        self.data[filename] = {
            'processed_at': datetime.now().isoformat(),
            'product_id': product_id,
            'storage_path': storage_path,
            'url': url
        }
        self.save()

processed = ProcessedImages()

def get_mime_type(extension):
    """Get MIME type from file extension"""
    mime_types = {
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.png': 'image/png',
        '.gif': 'image/gif',
        '.webp': 'image/webp'
    }
    return mime_types.get(extension.lower(), 'application/octet-stream')

def upload_to_storage(file_path):
    """Upload image to Supabase Storage"""
    file_name = file_path.name
    file_ext = file_path.suffix
    timestamp = int(time.time() * 1000)
    storage_path = f"{timestamp}{file_ext}"
    
    print(f"   ⬆️  Uploading to Supabase Storage...")
    
    with open(file_path, 'rb') as f:
        file_data = f.read()
    
    headers = {
        'apikey': SUPABASE_ANON_KEY,
        'Authorization': f'Bearer {SUPABASE_ANON_KEY}',
        'Content-Type': get_mime_type(file_ext)
    }
    
    url = f"{SUPABASE_URL}/storage/v1/object/products/{storage_path}"
    
    response = requests.post(url, headers=headers, data=file_data)
    
    if response.status_code in [200, 201]:
        public_url = f"{SUPABASE_URL}/storage/v1/object/public/products/{storage_path}"
        print(f"   ✅ Uploaded to: {public_url}")
        return storage_path, public_url
    else:
        raise Exception(f"Upload failed: {response.status_code} - {response.text}")

def create_product(image_path, file_name):
    """Create product in Supabase database"""
    # Generate product name from filename
    product_name = file_name.stem.replace('-', ' ').replace('_', ' ').title()
    
    print(f"   📦 Creating product in database...")
    
    product_data = {
        'name': product_name,
        'description': 'Producto pendiente de descripción - Editar en Supabase',
        'price': 0,
        'category': 'vestments',  # Default category
        'image': image_path
    }
    
    headers = {
        'apikey': SUPABASE_ANON_KEY,
        'Authorization': f'Bearer {SUPABASE_ANON_KEY}',
        'Content-Type': 'application/json',
        'Prefer': 'return=representation'
    }
    
    url = f"{SUPABASE_URL}/rest/v1/products"
    
    response = requests.post(url, headers=headers, json=product_data)
    
    if response.status_code == 201:
        product = response.json()[0]
        print(f"   ✅ Product created with ID: {product['id']}")
        return product
    else:
        raise Exception(f"Product creation failed: {response.status_code} - {response.text}")

def process_image(file_path):
    """Process a new image file"""
    file_name = file_path.name
    
    print(f"\n📸 Processing: {file_name}")
    
    try:
        # Upload to Storage
        storage_path, public_url = upload_to_storage(file_path)
        
        # Create product
        product = create_product(storage_path, file_path)
        
        # Mark as processed
        processed.mark_processed(file_name, product['id'], storage_path, public_url)
        
        print(f"   💡 Edit product details at:")
        print(f"      https://supabase.com/dashboard/project/nzwtafacdpdgulzcwntx/editor")
        print(f"   🎉 Success! Product '{product['name']}' ready for editing\n")
        
    except Exception as e:
        print(f"   ❌ Error: {str(e)}\n")

class ImageHandler(FileSystemEventHandler):
    """Handle file system events"""
    
    def on_created(self, event):
        if event.is_directory:
            return
        
        file_path = Path(event.src_path)
        
        # Check if it's a supported image format
        if file_path.suffix.lower() in SUPPORTED_FORMATS:
            # Check if not already processed
            if not processed.is_processed(file_path.name):
                # Wait a bit to ensure file is fully written
                time.sleep(1)
                if file_path.exists():
                    process_image(file_path)

def process_existing_images():
    """Process any existing unprocessed images"""
    if not IMAGES_FOLDER.exists():
        IMAGES_FOLDER.mkdir()
        print("✅ Created product-images folder\n")
        return
    
    image_files = [
        f for f in IMAGES_FOLDER.iterdir()
        if f.is_file() and f.suffix.lower() in SUPPORTED_FORMATS
        and not processed.is_processed(f.name)
    ]
    
    if image_files:
        print(f"📂 Found {len(image_files)} unprocessed image(s)\n")
        for image_file in image_files:
            process_image(image_file)

def main():
    """Main function"""
    print("=" * 60)
    print("🏛️  DIVINA PROVIDENTIA - Auto Upload System")
    print("=" * 60)
    print(f"\n👀 Watching folder: {IMAGES_FOLDER}")
    print(f"📝 Supported formats: {', '.join(SUPPORTED_FORMATS)}\n")
    
    # Process existing images
    process_existing_images()
    
    # Start watching for new images
    event_handler = ImageHandler()
    observer = Observer()
    observer.schedule(event_handler, str(IMAGES_FOLDER), recursive=False)
    observer.start()
    
    print("✨ Ready! Drop images in the product-images folder to auto-upload")
    print("   Press Ctrl+C to stop\n")
    
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\n\n👋 Stopping auto-upload system...")
        observer.stop()
    
    observer.join()
    print("✅ Stopped\n")

if __name__ == "__main__":
    main()

File: auto-upload.js
// Auto-Upload Product Images to Supabase
// This script watches the 'product-images' folder and automatically uploads new images

const SUPABASE_URL = 'https://nzwtafacdpdgulzcwntx.supabase.co';
const SUPABASE_SERVICE_KEY = 'YOUR_SERVICE_ROLE_KEY_HERE'; // You'll need to add this

const fs = require('fs');
const path = require('path');
const https = require('https');

const IMAGES_FOLDER = path.join(__dirname, 'product-images');
const PROCESSED_FILE = path.join(__dirname, '.processed-images.json');

// Load processed images history
let processedImages = {};
try {
    if (fs.existsSync(PROCESSED_FILE)) {
        processedImages = JSON.parse(fs.readFileSync(PROCESSED_FILE, 'utf8'));
    }
} catch (error) {
    console.error('Error loading processed images:', error);
}

// Save processed images history
function saveProcessedImages() {
    fs.writeFileSync(PROCESSED_FILE, JSON.stringify(processedImages, null, 2));
}

// Upload image to Supabase Storage
async function uploadToSupabase(filePath, fileName) {
    return new Promise((resolve, reject) => {
        const fileBuffer = fs.readFileSync(filePath);
        const fileExt = path.extname(fileName);
        const timestamp = Date.now();
        const storagePath = `${timestamp}${fileExt}`;

        const options = {
            hostname: 'nzwtafacdpdgulzcwntx.supabase.co',
            port: 443,
            path: `/storage/v1/object/products/${storagePath}`,
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
                'Content-Type': getMimeType(fileExt),
                'Content-Length': fileBuffer.length
            }
        };

        const req = https.request(options, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => {
                if (res.statusCode === 200 || res.statusCode === 201) {
                    resolve({
                        path: storagePath,
                        url: `${SUPABASE_URL}/storage/v1/object/public/products/${storagePath}`
                    });
                } else {
                    reject(new Error(`Upload failed: ${res.statusCode} - ${data}`));
                }
            });
        });

        req.on('error', reject);
        req.write(fileBuffer);
        req.end();
    });
}

// Create product in Supabase
async function createProduct(imagePath, fileName) {
    return new Promise((resolve, reject) => {
        const productName = path.basename(fileName, path.extname(fileName))
            .replace(/-/g, ' ')
            .replace(/_/g, ' ')
            .split(' ')
            .map(word => word.charAt(0).toUpperCase() + word.slice(1))
            .join(' ');

        const productData = {
            name: productName,
            description: 'Producto pendiente de descripción',
            price: 0,
            category: 'vestments', // Default category
            image: imagePath
        };

        const data = JSON.stringify(productData);

        const options = {
            hostname: 'nzwtafacdpdgulzcwntx.supabase.co',
            port: 443,
            path: '/rest/v1/products',
            method: 'POST',
            headers: {
                'apikey': SUPABASE_SERVICE_KEY,
                'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
                'Content-Type': 'application/json',
                'Prefer': 'return=representation'
            }
        };

        const req = https.request(options, (res) => {
            let responseData = '';
            res.on('data', (chunk) => responseData += chunk);
            res.on('end', () => {
                if (res.statusCode === 201) {
                    resolve(JSON.parse(responseData)[0]);
                } else {
                    reject(new Error(`Product creation failed: ${res.statusCode} - ${responseData}`));
                }
            });
        });

        req.on('error', reject);
        req.write(data);
        req.end();
    });
}

// Get MIME type from extension
function getMimeType(ext) {
    const types = {
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.png': 'image/png',
        '.gif': 'image/gif',
        '.webp': 'image/webp'
    };
    return types[ext.toLowerCase()] || 'application/octet-stream';
}

// Process a new image
async function processImage(fileName) {
    const filePath = path.join(IMAGES_FOLDER, fileName);

    console.log(`\n📸 Processing: ${fileName}`);

    try {
        // Upload to Supabase Storage
        console.log('   ⬆️  Uploading to Supabase Storage...');
        const { path: storagePath, url } = await uploadToSupabase(filePath, fileName);
        console.log('   ✅ Uploaded to:', url);

        // Create product in database
        console.log('   📦 Creating product in database...');
        const product = await createProduct(storagePath, fileName);
        console.log('   ✅ Product created with ID:', product.id);
        console.log('   💡 Edit details at: https://supabase.com/dashboard/project/nzwtafacdpdgulzcwntx/editor');

        // Mark as processed
        processedImages[fileName] = {
            processedAt: new Date().toISOString(),
            productId: product.id,
            storagePath,
            url
        };
        saveProcessedImages();

        console.log(`   🎉 Success! Product "${product.name}" ready for editing\n`);
    } catch (error) {
        console.error(`   ❌ Error processing ${fileName}:`, error.message);
    }
}

// Watch folder for new images
function watchFolder() {
    console.log('👀 Watching folder:', IMAGES_FOLDER);
    console.log('📝 Supported formats: .jpg, .jpeg, .png, .gif, .webp\n');

    // Create folder if it doesn't exist
    if (!fs.existsSync(IMAGES_FOLDER)) {
        fs.mkdirSync(IMAGES_FOLDER);
        console.log('✅ Created product-images folder\n');
    }

    // Process existing images
    const existingFiles = fs.readdirSync(IMAGES_FOLDER);
    const imageFiles = existingFiles.filter(file =>
        /\.(jpg|jpeg|png|gif|webp)$/i.test(file) && !processedImages[file]
    );

    if (imageFiles.length > 0) {
        console.log(`📂 Found ${imageFiles.length} unprocessed image(s)\n`);
        imageFiles.forEach(processImage);
    }

    // Watch for new files
    fs.watch(IMAGES_FOLDER, (eventType, fileName) => {
        if (eventType === 'rename' && fileName && /\.(jpg|jpeg|png|gif|webp)$/i.test(fileName)) {
            const filePath = path.join(IMAGES_FOLDER, fileName);

            // Check if file exists (added, not deleted) and not already processed
            if (fs.existsSync(filePath) && !processedImages[fileName]) {
                // Wait a bit to ensure file is fully written
                setTimeout(() => {
                    if (fs.existsSync(filePath)) {
                        processImage(fileName);
                    }
                }, 1000);
            }
        }
    });

    console.log('✨ Ready! Drop images in the product-images folder to auto-upload\n');
}

// Start watching
watchFolder();
